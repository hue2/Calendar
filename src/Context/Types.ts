export interface IEventDb {
    _id: string,
    note: string, 
    date: string,
    sticky: boolean,
}

export interface IContextProps {
    show: boolean,
    date: string,
    note: string,
    events: Array<IEvent>,
    importDest: string,
    sticky: boolean,
    stickyEvents: Array<IEvent>,
    onShow: () => void,
    onClose: () => void,
    onDatePick: (event : any) => void,
    onNoteSet: (event : any) => void,
    onEventSave: (id: string | null) => void,
    onEdit: (event : IEvent) => void,
    onDelete: () => void,
    onExportSelect: (data: any) => void,
    onExport: () => void,
    onImportSel: (data: string) => void,
    onImport: (event: any) => void;
    onSetSticky: (event: any) => void;
}

export interface IEvent {
    id: string,
    start: string,
    end: string,
    title: string,
    sticky: boolean
}
