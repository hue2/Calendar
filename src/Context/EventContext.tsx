import React, { useState, createContext, useEffect } from "react";

import Db from '../Database/Db';
import { IEventDb, IContextProps, IEvent } from './Types';

const fs = require('fs');

export const EventContext = createContext({} as IContextProps);

const db = new Db();

export function EventProvider(props: any) {
    let defaultDate = new Date().toISOString().substring(0, 10);

    const [show, setShow] = useState(false);
    const [date, setDate] = useState(defaultDate);
    const [note, setNote] = useState("");
    const [id, setId] = useState("");
    const [sticky, setSticky] = useState(false);
    const [events, setEvent] = useState<Array<IEvent>>([]);
    const [stickyEvents, setStickyEvents] = useState<Array<IEvent>>([]);

    useEffect(() => {
       getEvents();
    }, []);

    function handleShow() {
        setShow(true);
    }

    function getEvents() {
        db.getEvent().then((docs : any) => {
            let nonStickEvents = docs.filter(x => x.sticky === undefined || x.sticky === false);
            let allEvents = nonStickEvents.map((item : any) => ({
                id: item._id,
                title: item.note,
                start: item.date,
                end: item.end,
                sticky: false
            }));
            setEvent(allEvents);

            let stickEventsOnly = docs.filter(x => x.sticky !== undefined && x.sticky === true);
            let stickyEvents = stickEventsOnly.map((item : any) => ({
                id: item._id,
                title: item.note,
                start: item.date,
                end: item.end,
                sticky: true
            }));

            setStickyEvents(stickyEvents);
        });
    }

    function handleEditModalShow(event : IEvent) {
        setId(event.id);
        setNote(event.title);
        setDate(event.start);
        setSticky(event.sticky)
        setShow(true);
    }
    
    function handleClose() {
        handleReset();
        setShow(false);
    }

    function handleReset() {
        setId("");
        setNote("");;
        setDate(defaultDate);
        setSticky(false);
    }

    function handleDatePick(event : any) {
        setDate(event.target.value);
    }

    function handleNoteSet(event : any) {
        setNote(event.target.value);
    }
    
    function handleSetSticky(event : any) {
        setSticky(event.target.checked);
    }
    
    function handleEventSave() {
       if (id === null || id === "") {
            handleCreate();
       }
       else {
            handleEdit();
       }
    }

    function handleDelete() {
        db.deleteEvent(id).then(() => {
            handleClose();
            getEvents();
        })
    }

    function handleCreate() {
        db.createEvent({ date, note, sticky }).then((result : IEventDb) => {
            getEvents();
            handleClose();
        });
    }
     
    function handleEdit() {
        db.editEvent(id, { date, note, sticky }).then((result : IEventDb) => {
            getEvents();
            handleClose();
        });
    }  

    return (
        <EventContext.Provider value={{
                show, 
                date, 
                note,
                events,
                sticky,
                stickyEvents,
                onShow: handleShow, 
                onClose: handleClose,
                onDatePick: handleDatePick, 
                onNoteSet: handleNoteSet,
                onEventSave: handleEventSave,
                onEdit: handleEditModalShow,
                onDelete: handleDelete,
                onSetSticky: handleSetSticky,
            }}
        >
            {props.children}
        </EventContext.Provider>
    )
}
