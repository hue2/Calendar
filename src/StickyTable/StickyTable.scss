#sticky {
    text-align: left;
    color: white;
}

tbody > tr {
    &:hover {
        background: white;
        color: black;
    }
}

thead > tr {
    background: #306a87;
    &:hover {
       color: white !important; 
    }
}

td {
    white-space:pre;
}