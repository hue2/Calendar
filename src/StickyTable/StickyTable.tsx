import React from 'react';
import { Table } from 'react-bootstrap';

import "./StickyTable.scss";

export function StickyTable(props) {
    const { stickyEvents, handleEditModal } = props;

    return (
        <div data-testid="sticky-table">
            <Table id="sticky">
                <thead>
                <tr>
                    <th>Sticky Note</th>
                </tr>
                </thead>
                <tbody>
                {stickyEvents.length > 0 && stickyEvents.map(item => 
                    <tr onClick={() => handleEditModal(item)} key={item.id}>
                        <td>{item.title}</td>                 
                    </tr>
                )}
                </tbody>
            </Table>
        </div>
    )
}