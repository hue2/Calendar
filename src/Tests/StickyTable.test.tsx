import React from 'react';
import { render, getByTestId  } from '@testing-library/react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow } from 'enzyme';

import { EventContext } from '../Context/EventContext';
import { StickyTable } from '../StickyTable/StickyTable';

Enzyme.configure({ adapter: new Adapter() });

describe("test sticky table modal", () => {
    let contextCallback, stickyEvents, wrapper;

    beforeEach(() => { 
        stickyEvents = [
            {
                id: 1,
                title: 'hello',
            }
        ]
        contextCallback = jest.fn();
        wrapper = render(
            <EventContext.Provider value={contextCallback}>
                <StickyTable stickyEvents={stickyEvents}/>     
            </EventContext.Provider>,
        );
    });

    it("should match snapshot", () => {
        const tree = shallow(<StickyTable stickyEvents={stickyEvents} />);
        expect(tree).toMatchSnapshot();
    });

    it('should render', () => {
        const { getByTestId } =  wrapper;

        const table = getByTestId('sticky-table');
      
        expect(table).toBeInTheDocument();      
    })
})