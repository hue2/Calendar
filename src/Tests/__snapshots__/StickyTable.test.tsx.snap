// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test sticky table modal should match snapshot 1`] = `ShallowWrapper {}`;
